File: hard/spiral_printing.rb
File.open(ARGV[0]).each_line do |line|
  s = line.split(';')
  row = s[0].to_i
  col = s[1].to_i
  t = s[2].split
  i = tn = tw = 0
  j = 1
  te = col - 1
  ts = row - 1
  print t[0]
  loop do
    while j <= te
      print ' ' + t[i * col + j]
      j += 1
    end
    j -= 1
    i += 1
    tn += 1
    break if i > ts
    while i <= ts
      print ' ' + t[i * col + j]
      i += 1
    end
    i -= 1
    j -= 1
    te -= 1
    break if j < tw
    while j >= tw
      print ' ' + t[i * col + j]
      j -= 1
    end
    j += 1
    i -= 1
    ts -= 1
    break if i < tn
    while i >= tn
      print ' ' + t[i * col + j]
      i -= 1
    end
    i += 1
    j += 1
    tw += 1
    break if j > te
  end
  puts
end

File: hard/spiral_printing.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp $line;
    my ( $row, $col, $s ) = split( /;/, $line );
    my @t = split( / /, $s );
    my ( $i, $tn, $tw, $j, $te, $ts ) = ( 0, 0, 0, 1, $col - 1, $row - 1 );
    print "$t[0]";
    while (1) {
        for ( ; $j <= $te ; $j++ ) {
            print " $t[$i * $col + $j]";
        }
        $j--;
        $i++;
        $tn++;
        last if ( $i > $ts );
        for ( ; $i <= $ts ; $i++ ) {
            print " $t[$i * $col + $j]";
        }
        $i--;
        $j--;
        $te--;
        last if ( $j < $tw );
        for ( ; $j >= $tw ; $j-- ) {
            print " $t[$i * $col + $j]";
        }
        $j++;
        $i--;
        $ts--;
        last if ( $i < $tn );
        for ( ; $i >= $tn ; $i-- ) {
            print " $t[$i * $col + $j]";
        }
        $i++;
        $j++;
        $tw++;
        last if ( $j > $te );
    }
    print "\n";
}
close(INFILE);
